File: frontend/src/api/delivery-settings.ts
import apiClient from "@/api/client";
import { parseApiContract } from "@/api/contract";
import { z } from "zod";

export interface DeliverySettingsResponse {
  email_alerts_enabled: boolean;
  sms_alerts_enabled: boolean;
  version: number;
  updated_at: string;
  warnings: string[];
}

export interface DeliverySettingsUpdatePayload {
  email_alerts_enabled?: boolean;
  sms_alerts_enabled?: boolean;
  expected_version?: number;
}

const deliverySettingsResponseSchema: z.ZodType<DeliverySettingsResponse> = z
  .looseObject({
    email_alerts_enabled: z.boolean(),
    sms_alerts_enabled: z.boolean(),
    version: z.number(),
    updated_at: z.string(),
    warnings: z.array(z.string()),
  });

export const getMyDeliverySettings =
  async (): Promise<DeliverySettingsResponse> => {
    const response = await apiClient.get<DeliverySettingsResponse>(
      "/delivery-settings/me",
    );
    return parseApiContract(
      deliverySettingsResponseSchema,
      response.data,
      "/delivery-settings/me",
    );
  };

export const updateMyDeliverySettings = async (
  payload: DeliverySettingsUpdatePayload,
): Promise<DeliverySettingsResponse> => {
  const response = await apiClient.patch<DeliverySettingsResponse>(
    "/delivery-settings/me",
    payload,
  );
  return parseApiContract(
    deliverySettingsResponseSchema,
    response.data,
    "/delivery-settings/me",
  );
};

File: backend/app/api/v1/delivery_settings.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_active_user, get_db
from app.models.user import User
from app.schemas.delivery_settings import DeliverySettingsResponse, DeliverySettingsUpdateRequest
from app.services.delivery_settings_service import DeliverySettingsService

router = APIRouter(prefix="/delivery-settings", tags=["delivery-settings"])


def _require_advisor(current_user: User) -> None:
    if current_user.role != "advisor":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Advisor access required",
        )


@router.get(
    "/me",
    response_model=DeliverySettingsResponse,
    summary="Get delivery settings for current advisor",
)
def get_my_delivery_settings(
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> DeliverySettingsResponse:
    _require_advisor(current_user)
    settings = DeliverySettingsService.get_or_create_for_user(db=db, user_id=current_user.id)
    warnings = DeliverySettingsService.get_warnings_for_user(current_user, settings)
    return DeliverySettingsResponse(
        email_alerts_enabled=settings.email_alerts_enabled,
        sms_alerts_enabled=settings.sms_alerts_enabled,
        version=int(settings.version),
        updated_at=settings.updated_at,
        warnings=warnings,
    )


@router.patch(
    "/me",
    response_model=DeliverySettingsResponse,
    summary="Update delivery settings for current advisor",
)
def update_my_delivery_settings(
    payload: DeliverySettingsUpdateRequest,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db),
) -> DeliverySettingsResponse:
    _require_advisor(current_user)
    settings = DeliverySettingsService.update_for_user(
        db=db,
        user_id=current_user.id,
        email_alerts_enabled=payload.email_alerts_enabled,
        sms_alerts_enabled=payload.sms_alerts_enabled,
        expected_version=payload.expected_version,
    )
    warnings = DeliverySettingsService.get_warnings_for_user(current_user, settings)
    return DeliverySettingsResponse(
        email_alerts_enabled=settings.email_alerts_enabled,
        sms_alerts_enabled=settings.sms_alerts_enabled,
        version=int(settings.version),
        updated_at=settings.updated_at,
        warnings=warnings,
    )
